Tests/AsyncQueueTests/FIFOQueueTests.swift

```diff
@@ -65,7 +91,54 @@ final class FIFOQueueTests: XCTestCase {
         await systemUnderTest.await { /* Drain the queue */ }
     }
 
+    func test_asyncOn_executesAsyncBlocksAtomically() async {
+        let semaphore = Semaphore()
+        for _ in 1...1_000 {
+            systemUnderTest.async(on: semaphore) { semaphore in
+                let isWaiting = semaphore.isWaiting
+                // This test will fail occasionally if we aren't executing atomically.
+                // You can prove this to yourself by replacing `systemUnderTest.async` above with `Task`.
+                XCTAssertFalse(isWaiting)
+                // Signal the semaphore before or after we wait – let the scheduler decide.
+                Task {
+                    semaphore.signal()
+                }
+                // Wait for the concurrent task to complete.
+                await semaphore.wait()
+            }
+        }
+        await systemUnderTest.await { /* Drain the queue */ }
+    }
+
     func test_async_isNotReentrant() async {
+        let counter = Counter()
+        systemUnderTest.async { [systemUnderTest] in
+            systemUnderTest.async {
+                await counter.incrementAndExpectCount(equals: 2)
+            }
+            await counter.incrementAndExpectCount(equals: 1)
+            systemUnderTest.async {
+                await counter.incrementAndExpectCount(equals: 3)
+            }
+        }
+        await systemUnderTest.await { /* Drain the queue */ }
+    }
+
+    func test_asyncOn_isNotReentrant() async {
+        let counter = Counter()
+        systemUnderTest.async(on: counter) { [systemUnderTest] counter in
+            systemUnderTest.async(on: counter) { counter in
+                counter.incrementAndExpectCount(equals: 2)
+            }
+            counter.incrementAndExpectCount(equals: 1)
+            systemUnderTest.async(on: counter) { counter in
+                counter.incrementAndExpectCount(equals: 3)
+            }
+        }
+        await systemUnderTest.await { /* Drain the queue */ }
+    }
+
+    func test_await_async_areNotReentrant() async {
```

Comment:
this test was previously called `test_async_isNotReentrant`, but technically it was testing that `await` and `async` aren't reentrant rather than just testing that `async` is not reentrant.